kubeadm: Stop assuming full ownership of /etc/kubernetes.

dgoodwin opened the pull request:
Packages may auto-create directories in /etc/kubernetes, and users also
need files such as cloud-config.json to be present and preserved at
their default locations in /etc/kubernetes. As such this modifies
pre-flight checks to only require the absence of the files and
directories we explicitly create in kubeadm.

Reset is similarly modified to not wipe out /etc/kubernetes entirely.
When resetting directories we also now preserve the directory itself,
but delete it's contents.

Also adds tests for reset command logic specifically for /etc/kubernetes
cleanup, to ensure user files are not inadvertently wiped out.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35632)

<!-- Reviewable:end -->

dgoodwin committed:
kubeadm: Stop assuming full ownership of /etc/kubernetes.

Packages may auto-create directories in /etc/kubernetes, and users also
need files such as cloud-config.json to be present and preserved at
their default locations in /etc/kubernetes. As such this modifies
pre-flight checks to only require the absence of the files and
directories we explicitly create in kubeadm.

Reset is similarly modified to not wipe out /etc/kubernetes entirely.
When resetting directories we also now preserve the directory itself,
but delete it's contents.

Also adds tests for reset command logic specifically for /etc/kubernetes
cleanup, to ensure user files are not inadvertently wiped out.

dgoodwin commented:
CC @luxas @kubernetes/sig-cluster-lifecycle

luxas commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -113,31 +113,44 @@ func (irc IsRootCheck) Check() (warnings, errors []error) {
+
+func (fac FileAvailableCheck) Check() (warnings, errors []error) {
+	errors = []error{}
+	if _, err := os.Stat(fac.path); err == nil {
```
I assume it should be `err != nil` here

dgoodwin replied:
os.Stat actually throws an error if the file does not exist, so no error here means it does exist and thus we should fail the check. (alt example can be found in DirAvailableCheck)

luxas replied:
ping

luxas commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -113,31 +113,44 @@ func (irc IsRootCheck) Check() (warnings, errors []error) {
 
+// FileAvailableCheck checks that the given file does not already exist.
+type FileAvailableCheck struct {
+	path string
```
Set this to `Path` to be consistent with `DirAvailableCheck`

luxas replied:
ping

pires commented:
Glad to see tests coming in, great work @dgoodwin! I did a quick pass on this and it LGTM.

luxas commented on cmd/kubeadm/app/cmd/reset.go:
```diff
@@ -66,6 +67,58 @@ func NewReset(skipPreFlight bool) (*Reset, error) {
+
+	d, err := os.Open(path)
+	if err != nil {
+		fmt.Printf("failed to remove directory: [%v]\n", err)
```
`failed to remove contents of directory` would be better I think

dgoodwin commented:
I'm wondering now if I should revert to wiping our the dirs we do plan to clean out so there's less trace left on system? Right now if you reset, /var/lib/etcd for example remains, it's just empty. It's a little weird in the /etc/kubernetes/manifests case as that does come from the rpm, and maybe we should change that, but it's probably safe to remove anyhow. I think I might do this, reset should probably leave as little footprint as possible.

luxas commented on cmd/kubeadm/app/cmd/reset.go:
```diff
@@ -81,13 +134,12 @@ func (r *Reset) Run(out io.Writer) error {
+	resetConfigDir("/etc/kubernetes/")
+
+	dirsToClean := []string{"/var/lib/kubelet", "/var/lib/etcd"}
+	fmt.Printf("Deleting contents of stateful directories: %v\n", dirsToClean)
```
nit: if you could move this above `resetConfigDir("/etc/kubernetes/")` it would be more logical with potential errors etc.

luxas commented on cmd/kubeadm/app/cmd/reset.go:
```diff
@@ -66,6 +67,58 @@ func NewReset(skipPreFlight bool) (*Reset, error) {
+		if _, err := os.Stat(path); os.IsNotExist(err) {
+			continue
+		}
+		err := os.Remove(path)
```
`os.Remove` should equal `rm -r` and `os.RemoveAll` should equal `rm -rf` which doesn't complain about non-existant dirs/files so if you change to that you can remove the check above I think

luxas replied:
ping. what do you think about this?

luxas commented on cmd/kubeadm/app/cmd/reset.go:
```diff
@@ -81,13 +134,12 @@ func (r *Reset) Run(out io.Writer) error {
+	dirsToClean := []string{"/var/lib/kubelet", "/var/lib/etcd"}
+	fmt.Printf("Deleting contents of stateful directories: %v\n", dirsToClean)
+	for _, dir := range dirsToClean {
+		cleanDir(dir)
```
I think that we might want to just do a `os.RemoveAll` here to minimize leftover

luxas commented on cmd/kubeadm/app/cmd/reset.go:
```diff
@@ -66,6 +67,58 @@ func NewReset(skipPreFlight bool) (*Reset, error) {
+	}
+	fmt.Printf("Deleting contents of config directories: %v\n", dirsToClean)
+	for _, dir := range dirsToClean {
+		cleanDir(dir)
```
I think that we might want to just do a os.RemoveAll here to minimize leftover

If you're doing that, you can remove the cleanDir function which means less code => better

luxas reviewed (commented):
Overall, great!
Some nits and small comments that needs fixing, but nice to see tests here as well!

dgoodwin commented:
PR updated, reverted back to completely removing /etc/kubernetes/manifests, /etc/kubernetes/pki, /var/lib/etcd and /var/lib/kubelet. 

/etc/kubernetes/ will however be preserved, as will any files within it.

luxas commented on cmd/kubeadm/app/cmd/reset.go:
```diff
@@ -81,9 +112,11 @@ func (r *Reset) Run(out io.Writer) error {
+	resetConfigDir("/etc/kubernetes/")
+
+	dirsToClean := []string{"/var/lib/kubelet", "/var/lib/etcd"}
+	fmt.Printf("Deleting stateful directories: %v\n", dirsToClean)
```
move this above resetConfigDir

dgoodwin replied:
Hmm just the print? The config dir cleanup prints it's own similar message. 

Or do you mean the entire cleanup of the /var/lib dirs should occur before config dir cleanup.

luxas reviewed (commented):
don't know if you missed some of my comments, PTAL

dgoodwin commented:
Updated with feedback from @luxas.

luxas commented:
Thanks you @dgoodwin!

LGTM

luxas commented:
@dgoodwin verify wants you to update bazel build files...

cc @mikedanese

dgoodwin commented:
OK I think I figured it out, updated now. There was an additional e2e test build file that bazel also wanted to update, I did not include that in commit but if it fails here again I will add it. It feels like something local at this point.

errordeveloper commented:
@dgoodwin per SELinux discussion yesterday, could we just empty `/var/lib/kubelet`?

dgoodwin commented:
@errordeveloper yeah I already flip flopped on this once but now I'm thinking you're right, we should probably just scrub that dir and keep it intact, leaving the door open for the packages to own it and control selinux context. I will bring that code back in, please hold.

dgoodwin committed:
kubeadm: Empty directories during reset, but do not delete them.

This will allow packages to maintain ownership of config and data
directories, which may carry selinux or other attributes that should be
preserved, but we do not wish to manage within kubeadm itself.

dgoodwin commented:
Code restored, I am keeping this as a separate commit, after two reversals it feels a likely candidate for needing to be reverted or understood in the future.

luxas commented:
LGTM, let's get this in now!

Thanks

krousey commented on cmd/kubeadm/app/cmd/reset_test.go:
```diff
@@ -0,0 +1,182 @@
+func assertExists(t *testing.T, path string) {
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		t.Errorf("file/dir does not exist error: %s", err)
+		t.Errorf("file/dir does not exist: %s", path)
```
These can be combined into a single message.

krousey commented on cmd/kubeadm/app/cmd/reset_test.go:
```diff
@@ -0,0 +1,182 @@
+		// Verify the files we cleanup implicitly in every test:
+		assertExists(t, tmpDir)
+		assertNotExists(t, filepath.Join(tmpDir, "admin.conf"))
+		assertNotExists(t, filepath.Join(tmpDir, "kubelet.conf"))
```
Could you append these to the relevant slices and just check them in the same loops?

dgoodwin replied:
Idea was these should *always* not exist, so it seemed to make sense to always verify that rather than require each test to say so.

errordeveloper commented:
Can someone make a PR to release tools repo and update kubeadm to latest revision?

krousey commented on cmd/kubeadm/app/cmd/reset_test.go:
```diff
@@ -0,0 +1,182 @@
+		resetConfigDir(tmpDir)
+
+		// Verify the files we cleanup implicitly in every test:
+		assertExists(t, tmpDir)
```
Assert helpers like this are neat, but they lose the line number information when they log errors. Could you instead write a helper function that does the test and returns the result, but leave the `t.Errorf` calls closer to the actual failure?

dgoodwin replied:
Was not aware of this thanks, will keep in mind in future.

luxas commented:
Sure, but we may have to wait a little bit before the binaries are available (probably until tomorrow)

krousey reviewed (commented):
Just some minor nits. Nothing important :smile:

krousey commented:
FYI my comments aren't blocking... just suggestions

luxas commented:
Yes, those can be fixed in a follow-up when someone is enhancing our tests
